onboarding: Add one-time modals to introduce views.

prakhar1144 opened the pull request:
* Commit 1: Show one-time modal to introduce inbox view.
* Commit 2: Show one-time modal to introduce recent conversations view.
* Commit 3: Don't show these modals to existing users.

Fixes: #29073 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details><summary>Inbox</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/43791564-7dae-4781-bf88-7716051ec8ab">
</img>
</details> 

<details><summary>Recent Conversation</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/bbe99ee2-5e44-416e-9eaa-7ba0aeba8732">
</img>
</details> 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->
---
- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Posted few [questions in CZO](https://chat.zulip.org/#narrow/stream/137-feedback/topic/user.20research.3A.20Inbox.20and.20messages.20getting.20marked.20as.20read/near/1751983)

prakhar1144 commented on zerver/lib/hotspots.py:
```diff
@@ -86,6 +86,9 @@ def to_dict(self) -> Dict[str, str]:
         name="inbox_modal",
     ),
+    OneTimeNotice(
+        name="recent_conversations_modal",
```
I was willing to name these `intro_recent_conversations_modal` & `intro_inbox_modal`. But the character limit is 30 characters in our model. Are the current names okaish ?

timabbott replied:
If it's annoying, we can just do a quick migration to remove or increase the character limit to 50? There's no good reason to have such a short character limit.

timabbott replied:
But I think maybe `intro_recent_view_modal` and `intro_inbox_view_modal` might be better names that fit into the character limit anyway?

timabbott commented on tools/lib/capitalization.py:
```diff
@@ -24,6 +24,7 @@
     r"ID",
     r"IDs",
     r"IP",
+    r"Inbox",
```
You shouldn't need to be adding these as linter exceptions; (1) they follow our linting rules, and (2) I don't think it works well for translators to have this as a variable passed around.

prakhar1144 replied:
`tools/check-capitalization` gives this error
```
Strings not properly capitalized in frontend:
The <b>Inbox</b> view provides an overview of your conversations with unread messages.
```

alya replied:
Hm, maybe it's best to follow our own rules and not capitalize those. :)

prakhar1144 replied:
We decided to use capitalization in https://github.com/zulip/zulip/pull/29192#discussion_r1520565943

timabbott commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,26 @@
+<p>
+    {{#tr}}
+        Click on a conversation you're interested in to view it.
+        You can return to your <b>{zulip_view}</b> when done to select the next conversation.
```
"You can return to this view using the left sidebar when done to select the next conversation." might be better language anyway while also getting rid of the problematic-for-translation variable view name in this line; @alya thoughts.

alya replied:
How about this?

> You can return here using the left sidebar or the **back** button to select the next conversation.

timabbott replied:
Works for me.

prakhar1144 replied:
Made the changes. See updated screenshot in PR description.

alya commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,26 @@
+<p>
+    {{#tr}}
+        Tip: <b>{zulip_view}</b> is your home view, so use Esc to come back any time.
+        [if home view and Esc navigation to home view is enabled]
```
Sure, that seems fine to me.

prakhar1144 replied:
Fixed this. See updated screenshot in PR description.

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments.
* Added a migration to not show these modals to existing users.

I have included the [headings as discussed](https://chat.zulip.org/#narrow/stream/137-feedback/topic/user.20research.3A.20Inbox.20and.20messages.20getting.20marked.20as.20read/near/1752641). If we don't want to keep these headings, we'll have to [take a bit different approach](https://chat.zulip.org/#narrow/stream/137-feedback/topic/user.20research.3A.20Inbox.20and.20messages.20getting.20marked.20as.20read/near/1751984).

alya commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,27 @@
+    {{#if (eq zulip_view "inbox")}}
+        {{#tr}}The <b>Inbox</b> view provides an overview of your conversations with unread messages.{{/tr}}
+    {{else if (eq zulip_view "recent_conversations")}}
+        {{#tr}}The <b>Recent conversations</b> view provides an overview of all the ongoing conversations.{{/tr}}
```
We can make Inbox/Recent be lower-case.

alya commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,27 @@
+    {{/if}}
+    {{#tr}}
+        In Zulip, a conversation is either a <z-link-direct-message>direct message</z-link-direct-message>
+        thread (one-on-one or with a group), or a <z-link-stream-topic>topic in a stream</z-link-stream-topic>.
```
Let's drop " (one-on-one or with a group)".

alya commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,27 @@
+<p>
+    {{#tr}}
+        Click on a conversation you're interested in to view it.
+        You can return here using the left sidebar or the <b>back</b> button to select the next conversation.
```
Let's change "You can return here..." to:

To return here, you can:
- Use <b>back</b> button in your browser or desktop app.
- Click {{icon} **Inbox** or {icon} **Recent conversations**} in the left sidebar. [I guess we probably do want to capitalize here to match the UI, as we do [in the help center](https://zulip.com/help/inbox)?]
- Use <b>Esc</b> to go to your home view. [Show {{#if current_home_view_and_escape_navigation_enabled}}; this replaces the tip.]

alya replied:
@timabbott does this look good to you?

timabbott replied:
Sure that works for me!

alya commented on web/src/recent_view_ui.js:
```diff
@@ -1070,6 +1074,24 @@ export function show() {
+        });
+        dialog_widget.launch({
+            html_heading: $t_html({defaultMessage: "Welcome to Recent conversations"}),
+            html_body,
```
Let's try this?

- Welcome to **recent conversations**!
- Welcome to your **inbox**!

prakhar1144 commented:
<details><summary>Inbox</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/c7257fde-42d0-4dcb-92c3-422d59efbbaa">
</img>
</details> 

<details><summary>Recent conversations</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/25da4ae9-e978-43fa-a00b-752537caba40">
</img>
</details>

prakhar1144 commented:
Working on fixing those icon alignment issue.

alya commented:
Let's style Esc as a keyboard shortcut here (with capitalization, I think).

Otherwise looks good to me! (I haven't tested.)

prakhar1144 commented:
<details><summary>Inbox</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/1ed5ab8b-d845-4c14-b56a-e48bcd43f467">
</img>
</details>

alya commented:
Hmm, that styling stands out way more than we want it to in this context. Can we try the styling from our popover menus?

![Screenshot 2024-03-14 at 11 04 49 AM](https://github.com/zulip/zulip/assets/2090066/04359152-9860-4d13-9c21-917f536af21e)

timabbott commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,38 @@
+        {{/if}}
+    </p>
+    <p>
+        {{t "Click on a conversation you're interested in to view it."}}
```
"Click on a conversation to view it." would be simpler and easier to parse.

timabbott commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,38 @@
+                        <li>{{#tr}}Use <kbd>Esc</kbd> to go to your home view.{{/tr}}</li>
+                    {{/if}}
+                </ul>
+            {{/inline}}
```
Why is this an `inline` block? I would have expected to just have "To return here, you can:" be on `tr` block, and then the rest of this just inside that.

prakhar1144 replied:
Yes, that'd be more simplistic.

timabbott commented on web/templates/introduce_zulip_view_modal.hbs:
```diff
@@ -0,0 +1,38 @@
+                                Click <z-icon-inbox></z-icon-inbox> <b>Inbox</b> in the left sidebar.
+                                {{#*inline "z-icon-inbox"}}<i class="zulip-icon zulip-icon-inbox" aria-hidden="true"></i>{{/inline}}
+                            {{/tr}}
+                        {{/if}}
```
Presumably we should have a block similar to this one for the recent view?

prakhar1144 commented:
> If that doesn't work well, maybe just putting `Esc` in a code span will be best.

<details><summary>Option 1</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/cfe628e6-c9e5-4420-9ac9-6434bd9c8298">
</img>
</details> 

<details><summary>Option 2</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/1fcac7e6-70ac-462f-8dc7-9bafc705d823">
</img>
</details> 

I think, option 2 is better ?

prakhar1144 commented:
<details><summary>Inbox</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/ad804ddb-bb8f-4eb5-9ecf-df229c7f1ca2">
</img>
</details> 

<details><summary>Recent Conversation</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/9ea35596-9a3b-485b-b47c-78bfe0c7ab14">
</img>
</details>

timabbott commented:
I think option 2 is better -- in part because the element isn't so big that it disrupts the bulleted list spacing.

timabbott commented:
This lgtm aside from the above and https://github.com/zulip/zulip/pull/29192#discussion_r1525327766, which can probably be fixed after merging if it's an issue, but I'll bump it in Zulip.

alexmv commented on zerver/migrations/0501_mark_introduce_zulip_view_modals_as_read.py:
```diff
@@ -0,0 +1,49 @@
+            OnboardingStep(user=user_profile, onboarding_step="intro_recent_view_modal")
+        )
+
+    OnboardingStep.objects.bulk_create(onboarding_steps, batch_size=10000)
```
[Answered on CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/onboarding_steps.20migration/near/1759136), but this would be very easy to write as a `INSERT INTO ... SELECT` statement, and would be faster and less memory-intensive.

prakhar1144 commented:
Updates in the recent push:
* https://github.com/zulip/zulip/pull/29192#discussion_r1526693807 (3rd commit)
* https://github.com/zulip/zulip/pull/29192#issuecomment-2000235732 Changes related to this:


```diff
diff --git a/web/styles/modal.css b/web/styles/modal.css
index 072ecee5d4..689ff2a00f 100644
--- a/web/styles/modal.css
+++ b/web/styles/modal.css
@@ -472,4 +472,13 @@
     i {
         vertical-align: middle;
     }
+
+    .keyboard-button {
+        color: var(--color-hotkey-hint);
+        font-size: 12px;
+        font-weight: 500;
+        padding: 2px 4px;
+        border-radius: 3px;
+        border: 1px solid var(--color-hotkey-hint);
+    }
 }
diff --git a/web/templates/introduce_zulip_view_modal.hbs b/web/templates/introduce_zulip_view_modal.hbs
index d85bec03d8..d6ae60a13a 100644
--- a/web/templates/introduce_zulip_view_modal.hbs
+++ b/web/templates/introduce_zulip_view_modal.hbs
@@ -26,7 +26,12 @@
                 {{/if}}
             </li>
             {{#if current_home_view_and_escape_navigation_enabled}}
-                <li>{{#tr}}Use <kbd>Esc</kbd> to go to your home view.{{/tr}}</li>
+                <li>
+                    {{#tr}}
+                        Use <z-button>Esc</z-button> to go to your home view.
+                        {{#*inline "z-button"}}<span class="keyboard-button">{{@partial-block}}</span>{{/inline}}
+                    {{/tr}}
+                </li>
             {{/if}}
         </ul>
     </p>
```

prakhar1144 commented:
<details><summary>Inbox view</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/0cf3daba-48cd-4ee3-a109-0b1702a9ea5e">
</img>
</details> 

<details><summary>Recent view</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/b7f692ad-c2c9-4cc3-a60d-846cb6d28aed">
</img>
</details>

timabbott commented on zerver/migrations/0501_mark_introduce_zulip_view_modals_as_read.py:
```diff
@@ -0,0 +1,73 @@
+                UNION ALL
+                SELECT id, 'intro_recent_view_modal', %(timestamp_value)s
+                FROM zerver_userprofile
+                WHERE is_bot = False
```
This should probably include `AND is_mirror_dummy = False` -- mirror dummies are usually imported users who haven't used Zulip, and should be eligible for these onboarding items.

prakhar1144 replied:
Done.

alexmv commented:
My comments ended up here: https://github.com/zulip/zulip/commit/b821ba7c9be380bf460aef0c3ff4ac12143eaa0a#r139922730

prakhar1144 commented:
I have addressed the review comment & changes Alex suggested. -- 3rd commit.

timabbott commented:
This looks great, merged, thanks @prakhar1144!
